### PR TITLE
Selective AsRef/AsMut impl

### LINKED ISSUE #285: AsRef restricted impls
Currently there are 2 options for `AsRef` derives:

- the default behaviour, where only implementations for field types are generated
- `[as_ref(forward)]`, which generates a generic impl for all types the field can be converted to

It would be useful to be able to list types for implementation explicitly, like with [From](https://docs.rs/derive_more/1.0.0-beta.3/derive_more/derive.From.html)

For example

```rust
#[derive(AsRef)]
struct Test {
    #[as_ref(str)]
    str_field: String,
    #[as_ref([u8])]
    u8_field: Vec<u8>,
}
```

This wouldn't work with `[as_ref(forward)]`, since we'd have conflicting implementations for `AsRef<[u8]>`.

### PR DESCRIPTION
**Pull request by MegaBluejay**
Resolves #285 

## Synopsis

The `AsRef`/`AsMut` derives are less powerful than `From` and `Into`, because there's no way to specify types into which forwarded implementations should convert.

## Solution

Keep using the `forward` attribute argument for blanket forwarded impls, and use a list of types for selective ones.

## Checklist

- [x] Documentation is updated (if required)
- [x] Tests are added/updated (if required)
- [x] [CHANGELOG entry][l:1] is added (if required)

[l:1]: /CHANGELOG.md

### CODE REVIEW

File: impl/src/as/mod.rs [hunk 1]
```diff
@@ -178,92 +184,120 @@ impl<'a> ToTokens for Expansion<'a> {
...
+///
+/// ```rust,ignore
+/// #[as_ref(forward)]
+/// #[as_ref(<types>)]x
+/// ```
+enum AsArgs {
+    /// Blanket impl, fully forwarding to the field type
+    Forward(forward::Attribute),
+    /// Forward implementation, but only impl for specified types
+    Types(Punctuated<Type, Token![,]>),
```
Comment (MegaBluejay) [inline impl/src/as/mod.rs:285]: I don't think it makes sense to extract the type list attribute into `utils` like forward/skip, even though
the same arguments are used in `From` and `Into`, because parsing it is quite concise anyway

**Review by tyranron — CHANGES_REQUESTED** [review]
@MegaBluejay 

> I don't think it makes sense to extract the type list attribute into `utils` like forward/skip, even though
> the same arguments are used in `From` and `Into`, because parsing `Punctuated` is quite concise and understandable already

I don't really like this, there are still a lot of unnecessary duplication. But for keeping things simpler: refactoring this is beyond this PR's topic, and would be done in a separate now. So, for merging this PR, it's OK.

------------------------------------

Regarding the PR itself, it lacks implementation of one essential feature. See the comments below.

I think it was missed due to the fact, that you first wrote some code, and only latter added the tests, covering what you had implemented with your code. This led to the huge implementation-bias when you were writing the tests, as you thought only about what you had implemented.
That's why I propose and advise you to take the opposite, more TDDish, approach in the future, when implementing new features: first, you think of all feature combinations and edge cases, and fill them up as tests cases, so you have some high-level overview of what should be done, and cannot forget them, and only then you write the implementation covering those tests cases.

File: tests/as_ref.rs [hunk 2]
```diff
@@ -63,6 +66,51 @@ mod single_field {
             assert!(ptr::eq(rf, item.0.as_ref()));
         }
 
+        #[derive(AsRef)]
+        #[as_ref(i32, f64)]
+        struct Types(Foo);
+
+        // Asserts that the macro expansion doesn't generate `AsRef` impl for unmentioned type, by
+        // producing trait implementations conflict error during compilation, if it does.
+        impl AsRef<bool> for Types {
```
[the remark this answers is not in the document]
Comment (MegaBluejay) [reply tests/as_ref.rs:75]: The field type `Foo` implements `AsRef<bool>`, and
the assertion is there to make sure no blanket impl is generated.
I'll reword the comment to make it more clear what's going on

File: impl/doc/as_ref.md [hunk 3]
```diff
@@ -61,6 +61,37 @@ where
 }
 ```
 
+It's also possible to specify concrete types to derive forwarded
+impls for with `#[as_ref(<types>)]`.
```
[the remark this answers is not in the document]
Comment (tyranron) [reply impl/doc/as_ref.md:65]: Oh, and of course, we should have `compile_fail` tests covering this situation.

File: tests/as_ref.rs [hunk 4]
```diff
@@ -452,6 +681,28 @@ mod multi_field {
             assert!(ptr::eq(rf, item.first.as_ref()));
         }
 
+        #[derive(AsRef)]
+        struct Types {
+            #[as_ref(str)]
+            first: String,
```
[the remark this answers is not in the document]
Comment (MegaBluejay) [reply tests/as_ref.rs:862]: @tyranron 
I don't think this syntax can work on stable.

Autoref-based specialization gets most of the way there, but can't handle generics.

For instance, without something to differentiate these 2 cases, there's no way of generating correct impl bounds for both

```rust
// where T: TypeEquals<Other = T>
#[derive(AsRef)]
#[as_ref(T)]
struct Foo<T>(T);

// where T: AsRef<i32>
#[derive(AsRef)]
#[as_ref(i32)]
struct Bar<T>(T);
```

As an alternative, we could allow multiple attributes

```rust
#[derive(AsRef)]
#[as_ref]
#[as_ref(str)]
struct Foo(String);
```

Or use a special keyword for the field type

```rust
#[derive(AsRef)]
#[as_ref(str, direct)]
struct Foo(String);
```

Either one would remove the need for specialization.

[the remark this answers is not in the document]
Comment (tyranron) [reply tests/as_ref.rs:846]: @MegaBluejay 

> Autoref-based specialization gets most of the way there, but can't handle generics.

But do we need to?

> ```rust
> #[derive(AsRef)]
> #[as_ref(T)]
> struct Foo<T>(T);
> ```

This seems to be trivial, as we cannot alias a type parameter. So we can just check whether it contains in the type generics and treat this situation as `#[as_ref]`.

[the remark this answers is not in the document]
Comment (MegaBluejay) [reply tests/as_ref.rs:846]: One case I don't think we could handle this way is when the field type is itself generic and uses a type parameter

```rust
#[derive(AsRef)]
#[as_ref(i32)]
struct Foo<T>(T);

#[derive(AsRef)]
#[as_ref(Foo<T>, T)]
struct Bar<T>(Foo<T>);
```

`&Bar<T> -> &T` should be forwarded, even though there's a type parameter involved.

I'll look into this some more and figure out if there's a good way of working around this

[the remark this answers is not in the document]
Comment (tyranron) [reply tests/as_ref.rs:846]: @MegaBluejay can you show the expansion with autoref-based specialization for this case with the required types machinery?

[the remark this answers is not in the document]
Comment (MegaBluejay) [reply tests/as_ref.rs:846]: @tyranron [something like this](https://gist.github.com/MegaBluejay/945b19de936a76783b52bb29d4229082)

[the remark this answers is not in the document]
Comment (tyranron) [reply tests/as_ref.rs:846]: @MegaBluejay

First, I don't understand all the squats around `Cell`. Seems totally redundant and misleading, so I've removed it. It's not required for type-level reasoning (coherence) at all.

Second, your example seems to work just fine with autoref-based specialization, even when a type alias is inclined: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b54d0c62dea54208fe910fea5abfdb9f).

Third, we may omit using autoref-based specialization in the cases where the specified type is named exactly the same as the one of the field:
```rust
#[derive(AsRef)]
#[as_ref(Vec<T>)]
struct Stack<T>(Vec<T>);

// expands to just:
impl<T> AsRef<Vec<T>> for Stack<T> {
    fn as_ref(&self) -> &Vec<T> {
        &self.0
    }
}
```
We need only involve autoref-based specialization machinery when some "outer" types are used.

Given all of these, it seems that we're covering all the possible cases here. If not, prove me wrong with an example that doesn't compile on playground, please.

[the remark this answers is not in the document]
Comment (tyranron) [reply tests/as_ref.rs:846]: @MegaBluejay ah yes, of course, we cannot conditionally generate that bound:
```rust
where
    Foo<T>: AsRef<T>,
```
We either can generate it for both impls, which breaks the first impl, or doesn't generate it at all, which breaks the second impl.

Let me think over it a little bit... it seems that the problem is not fundamental, and we can overcome it by bounding with `Conv<_>: ExtractRef` somehow, which will automatically imply the necessary conditions.

[the remark this answers is not in the document]
Comment (tyranron) [reply tests/as_ref.rs:846]: @MegaBluejay 

> `Cell` is required for the `AsMut` version, because `Conv` needs to contain a mutable reference.

Yes, variance gives there some hard times for mutable references. Let's keep separate machinery traits for `AsRef` and `AsMut` versions, so we don't bother `AsRef` version with runtime impact of a `Cell`. Regarding `AsMut`, for now, use `Cell` in machinery, but give some more time poking around to make it working without `Cell`. There should be a way, imho.

> As for bounding with `Conv`, I don't think it's possible, since code using autoref-specialization isn't actually valid for all its cases at once, it just compiles differently in different contexts, which can't be expressed in types

True. We cannot express "OR" bounds, only "AND", so bringing all the bounds, we disjunct both cases, rather than conjunct them.

But we can use some trickery here, inspecting the type names. The whole algorithm would look like something like this:
1. We inspect the provided type to match by name the field type, if it does, we just use trivial `#[as_ref]` logic without specialization. It will easily handle all the cases of `impl<T> AsRef<T> for T`
    ```rust
    #[derive(AsRef)]
    #[as_ref(Vec<T>)]
    struct Stack<T>(Vec<T>);

    // expands to just:
    impl<T> AsRef<Vec<T>> for Stack<T> {
        fn as_ref(&self) -> &Vec<T> {
            &self.0
        }
    }
    ```
    except the cases where type is renamed
    ```rust
    type V<T> = Vec<T>;

    #[derive(AsRef)]
    #[as_ref(V<T>)]
    struct Stack<T>(Vec<T>);
   ```
2. If the provided type doesn't match the field type, we use the specialization wrapper.
3. If the provided type contains one of type parameters of the declared type (we can recursively inspect it), we add the `FieldType: AsRef<ProvidedType<T>>` bound, which enables the second specialization impl to work in generic contexts. Even more, such as we choose here the second specialization impl only, we may not use specialization totally here, falling into the assumption, that if the types named differently - they are different types.

This way, the only situations that don't work would be when we convert directly to the field type, which involves generics, and is named differently in field and attribute:
```rust
type V<T> = Vec<T>;

#[derive(AsRef)]
#[as_ref(V<T>)]
struct Stack<T>(Vec<T>);
```

I would say that this is a very rare situation. Why would one name differently __the  same type in the same place__? Moreover, we should state explicitly in the docs, that this concrete edge case won't work as expected.

Note, that we still have specialization working okay when generics are not involved, allowing to use renamed types.

-----------------------------------------

And another direction that is worth a shot, is another kind of specialization: runtime same-type specialization, which [inspects a type to be the same in runtime and transmutes](https://github.com/graphql-rust/juniper/blob/b375146466bbd19a3c4954e20d272e46842c04be/juniper/src/value/mod.rs#L145-L163), but `TypeId` and `any` machinery is [limited to `'static` lifetime only](https://doc.rust-lang.org/stable/std/any/struct.TypeId.html#method.of), unfortunately, which would likely limit kinds of types we can place the macro onto very-very unpractically much.

[the remark this answers is not in the document]
Comment (MegaBluejay) [reply tests/as_ref.rs:846]: I wrote out some more impls manually following this algorithm

We'd need to omit specialization in a wider number of cases, both when the provided type contains type parameters, and when the field type does.

For instance, this case

```rust
#[derive(AsRef)]
#[as_ref(i32)]
struct Foo<T>(T);
```

Specialization can't work here either, so we must assume `i32 != T` and use a forwarded impl.

Other than that everything seems to work.

**Review by tyranron — APPROVED** [review]
@MegaBluejay good job, thanks! 🍻

**Review by tyranron — APPROVED** [review]
@JelteF this one is relatively big and introduces a new feature. Though, I think it's polished quite well for now. Would you like to take a look?

**Comment by tyranron** [comment]
@JelteF ping

**Comment by JelteF** [comment]
I'll take a look this weekend. I'd like to double check if I like the the API interface

**Comment by tyranron** [comment]
@JelteF ping

**Review by JelteF — APPROVED** [review]
